core/src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java

```diff
@@ -207,4 +289,130 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
     public String getWriteableName() {
         return NAME;
     }
+
+    @Override
+    protected Query doToQuery(QueryShardContext context) throws IOException {
+        List<Object> terms = new ArrayList<>();
+        if (isTermsLookupQuery()) {
+            String lookupIndex = (this.lookupIndex == null) ? context.index().name() : this.lookupIndex;
+            handleTermsLookup(new TermsLookup(lookupIndex, lookupType, lookupId, lookupRouting, lookupPath, context.parseContext()), context.client(), terms);
+        } else {
+            terms.addAll(values);
+        }
+        if (terms.isEmpty()) {
+            return Queries.newMatchNoDocsQuery();
+        }
+        return handleTermsQuery(terms, fieldName, context, disableCoord);
+    }
+
+    static void handleTermsLookup(TermsLookup lookup, Client client, List<Object> terms) {
+        GetRequest getRequest = new GetRequest(lookup.getIndex(), lookup.getType(), lookup.getId()).preference("_local").routing(lookup.getRouting());
+        getRequest.copyContextAndHeadersFrom(SearchContext.current());
+        final GetResponse getResponse = client.get(getRequest).actionGet();
+        if (getResponse.isExists()) {
+            List<Object> extractedValues = XContentMapValues.extractRawValues(lookup.getPath(), getResponse.getSourceAsMap());
+            terms.addAll(extractedValues);
+        }
+    }
+
+    static Query handleTermsQuery(List<Object> terms, String fieldName, QueryShardContext context, boolean disableCoord) {
+        MappedFieldType fieldType = context.fieldMapper(fieldName);
+        String indexFieldName;
+        if (fieldType != null) {
+            indexFieldName = fieldType.names().indexName();
+        } else {
+            indexFieldName = fieldName;
+        }
+
+        Query query;
+        if (context.isFilter()) {  // NO COMMIT: not exactly sure how to handle this case, right now using a filter flag
```

Comment:
I think this is ok, the isFilter() flag should be set when parent queries that are already using toQuery() produce some inner query in a filter context.